### PR TITLE
feat/implement cancel option

### PR DESCRIPTION
**Pull request by shababali**
**Context**
Adding implementation and basic tests for the cancelOption functionality. The logic around seller cancelling an expired option threw me off, so Peter graciously helped clean my understanding and code blockers there (updated scenarios). Also, thanks again for the IDE setup and dev training! <3

I am putting this PR up in this state, because I don't want to slow down UI development. 
Sorry if the formatting wasn't up to par -- I was genuinely confused by what the expected convention/best practise was. Perhaps, we can go over this in standup?

**Notes**
Can add more tests especially in area of --
+ Eth transfer (re: seller cancels expired option)
+ inspecting properties of the option (struct) in cancelled (CLOSED) state

### CODE REVIEW

File: contracts/NFTOpt.sol [hunk 1]
```diff
@@ -184,10 +184,55 @@ contract NFTOpt {
         emit Filled(msg.sender, _optionId);
     }
 
-    function cancelOption(uint32 _optionId)
+    event Canceled(address, uint _optionId);  // TODO: move to top
+
+    // TODO:
+    // nice we have custom errors! (Thanks Luis -- are these easy to test though?)
+    // https://blog.soliditylang.org/2021/04/21/custom-errors/
```
Comment (shababali) [inline contracts/NFTOpt.sol:191]: old messaging to be cleaned

File: tsconfig.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "compilerOptions": {
-    "target": "ES2016",
+    "target": "ES2018",
```
Comment (shababali) [inline tsconfig.json:3]: The  change to `tsconfig.json` is to be able to learn Javascript/Typescript along w/ `hardhat.ethers` and `chai` in a Typescript REPL

File: test/NFTOpt.tests.ts [hunk 3]
```diff
@@ -286,9 +287,194 @@ describe("NFTOpt Tests", function () {
...
+
+            await NFTOptCTR.connect(seller).createOption(
+                1, {value: dummyOptionRequest.strikePrice}
+                );
+
+            const initialBuyerBalance = await buyer.getBalance();
+            const initialSellerBalance = await seller.getBalance();
+
+            // buyer cancels option
+            let tx = NFTOptCTR.connect(buyer).cancelOption(1);
```
Comment (Stefaanna) [inline test/NFTOpt.tests.ts:440]: what does `tx` mean?

File: contracts/NFTOpt.sol [hunk 4]
```diff
@@ -184,10 +184,55 @@ contract NFTOpt {
         emit Filled(msg.sender, _optionId);
     }
 
-    function cancelOption(uint32 _optionId)
+    event Canceled(address, uint _optionId);  // TODO: move to top
+
+    // TODO:
+    // nice we have custom errors! (Thanks Luis -- are these easy to test though?)
+    // https://blog.soliditylang.org/2021/04/21/custom-errors/
+    // implement in code
```
Comment (grasshopper47) [inline contracts/NFTOpt.sol:192]: remove comment

File: contracts/NFTOpt.sol (hunk 1 above)
Comment (grasshopper47) [reply contracts/NFTOpt.sol:191]: remove

File: contracts/NFTOpt.sol [hunk 5]
```diff
@@ -184,10 +184,55 @@ contract NFTOpt {
...
+            option.interval    == 0          &&
+            option.startDate   == 0          &&
+            option.premium     == 0          &&
+            option.strikePrice == 0
+        )
+        {
+            revert("The Option does not exist");
+        }
+
+        // cancel only filled or expired contracts
```
Comment (grasshopper47) [inline contracts/NFTOpt.sol:214]: remove comment

File: contracts/NFTOpt.sol [hunk 6]
```diff
@@ -184,10 +184,55 @@ contract NFTOpt {
...
+        if (expirationDate > block.timestamp)
+        {
+            require(msg.sender == option.buyer, "Only Buyer can cancel");
+        }
+        else
+        {
+            require(msg.sender == option.buyer || msg.sender == option.seller, "Only Buyer or Seller can cancel");
+        }
+
+        // transfer collateral from escrow to seller; update Option state to CLOSED
```
Comment (grasshopper47) [inline contracts/NFTOpt.sol:227]: remove comment

File: contracts/NFTOpt.sol [hunk 7]
```diff
@@ -184,10 +184,55 @@ contract NFTOpt {
...
+        (bool success,) = option.seller.call{value: option.strikePrice}("");
+        if (!success)
+        {
+            revert FUNDS_TRANSFER_FAILED();
+        }
+
+        options[_optionId].state = OptionState.CLOSED;
+        emit Canceled(msg.sender, _optionId);
```
Comment (grasshopper47) [inline contracts/NFTOpt.sol:236]: remove newline

File: test/NFTOpt.tests.ts [hunk 8]
```diff
@@ -286,9 +287,194 @@ describe("NFTOpt Tests", function () {
     });
 
     describe("cancelOption", function () {
+
         it("should test that method can be called", async function () {
```
Comment (grasshopper47) [inline test/NFTOpt.tests.ts:291]: remove these lines, the method can be called

File: test/NFTOpt.tests.ts [hunk 9]
```diff
@@ -286,9 +287,194 @@ describe("NFTOpt Tests", function () {
...
+        it("should test that the option exists", async function () {
+
+            expect(
+                NFTOptCTR.connect(buyer).cancelOption(0)
+                ).to.be.revertedWith("The Option does not exist");
+        });
+
+        it("should revert when the option does not have a start date i.e. published: REQUEST state", async function () {
+
+            await NFTOptCTR.connect(buyer).publishOptionRequest
```
Comment (grasshopper47) [inline test/NFTOpt.tests.ts:304]: there is a method wrapper for this called `publishOptionRequestDummy()` use that instead, it's not needed here to test an American option specifically

File: test/NFTOpt.tests.ts [hunk 10]
```diff
@@ -286,9 +287,194 @@ describe("NFTOpt Tests", function () {
...
+            ,   OptionFlavor.American
+            ,   {value: dummyOptionRequest.premium}
+            );
+
+            // filled
+            await NFTOptCTR.connect(seller).createOption(
+                1, {value: dummyOptionRequest.strikePrice}
+                );
+
+            var randomAccount = (await ethers.getSigners())[3];
```
Comment (grasshopper47) [inline test/NFTOpt.tests.ts:406]: create a global variable called account3 and use that

File: test/NFTOpt.tests.ts (hunk 3 above)
Comment (grasshopper47) [reply test/NFTOpt.tests.ts:440]: i think he copy/pasted your code :)

**Pull request closed without merging**